### ISSUE TITLE
Decide on a design for a minimum viable v0.1

### ISSUE THREAD

**Issue by ssokolow** [ssokolow/nodo#1]
As a "Just Works™ baseline sandboxing wrapper for any build automation", what features does this project need for a minimum viable product, and what features should be agreed on as definitely suitable for a later release?

## Current ideas for the design:

So far I'm thinking:

* To ensure that prepending this command doesn't changing the argument-parsing behaviour of the command being sandboxed, the command-line handling won't use an existing argument parser but, instead, will just check...
  * If there are no arguments, or there's only one argument and it's `-h` or `--help`, display help.
  * If the first argument is a valid profile name, launch the argument list (minus `argv[0]`), unmodified, as a command inside Firejail.
  * Otherwise, display a helpful "couldn't find a profile for that. See `--help` for more details." error and exit.
* Recommend the use of something like `alias nargo=nodo cargo` so that, if the alias is missing, they'll get an error message rather than an un-sandboxed execution.
* Try not to tie the sandboxing configuration to Firejail's CLI options specifically.

...and a configuration file design which provides this sort of functionality, but not necessarily with a schema anything like this:

```toml
# Default list of root-relative paths to be denied access to
# (The idea being to provide an analogue to `chattr +a foo.log`
# so `git diff` can be used to reveal shenanigans)
blacklist=[".git", ".hg", ".bzr", ".svn"]

[profile.cargo]
root_marked_by=["Cargo.toml"]
root_find_outermost=true  # For workspaces
projectless_subcommands=["init", "new"] # Assume $PWD is project root
allow_dbus_subcommands=["run"]
allow_gui_subcommands=["run"]
# allow_network=false
allow_network_subcommands=["add", "audit", "b", "build", "c", "check", "fetch", "run"] # etc. etc. etc.
deny_subcommands=["install", "uninstall"]  # must be run unconstrained

[profile.make]
root_marked_by=["Makefile"]
cwd_to_root=true  # run `make` in project root no matter where `nodo make` is run from
```

### Requirements for v0.1 MVP

* [ ] Settle on a name. `nodo` (Like "sudo" but "no") is a placeholder I'm hoping to replace because there are too many Google results for it already.
* [x] Identify which Firejail restrictions will be hard-coded for the v0.1 MVP and which ones should map to which config keys.
* [x] Decide whether to implement `root_find_outermost` for workspaces support for v0.1 or defer it to v0.2
* [x] ~~`cwd_to_root` to allow tools like `make` to handle being invoked from subdirectories~~
* [ ] Automatically blacklist all files that affect sandboxing behaviour
* [ ] A global `blacklist` config key to allow blacklisting things like `.git` so malicious packages can't hide their changes by amending un-pushed commits
* [ ] `root_marked_by` for automatically determining where to place the sandboxing boundary if something like `nodo cargo` or `nodo just` gets run in a subdirectory.
* [ ] `projectless_subcommands` to avoid forcing users to bypass the wrapper for `cargo new` and `cargo init`. (May need to redesign the actual config key for v0.2 to support things akin to `npm install -g` where it's not the subcommand that decides local vs. global)
* [ ] Decide on a redesign of `deny_subcommands` to support transparently allowing `npm install` but refusing with a "please bypass the wrapper" message for `npm install -g`.
* [ ] Decide what rules need to be in place to avoid pathological "ascend out of the project for lack of a `root_marked_by` file" situations. (Maybe stop and refuse if we hit `$HOME`, the home directory as identified by `getpwnam`/`getpwuid`, or a directory we lack write or execute/search permissions on the way up the tree before we find such a file.)
* [ ] Decide whether there are better names for any of these configuration keys

### v0.1.1

* [ ] `allow_network` and `allow_network_subcommands` (Just allowing network for all commands is an acceptable v0.1 as long as it still constrains filesystem access and blocks off D-Bus and the like)
* [ ] `root_resolve_symlinks=false` to allow using a differentially fuzzed port of Python's `os.path.abspath` instead of `std::fs::canonicalize`. (Needed to ensure we can have `root_marked_by` match the behaviour of all types of build automation if there's a symlink between `$PWD` and the root directory it finds.)
* [ ] Consider `allow_dbus_subcommands` and `allow_gui_subcommands` as a design for allowing things like `cargo run` access to D-Bus and X11 and the like.

### v0.1.x

* [ ] Brainstorm how much network whitelisting is viable with a "Just Works™ by default" config. (eg. What ports? Does crates.io's data schema allow it to arbitrarily point requests for bulk data at random CDN endpoints that are only known after Cargo has started? etc.)
* [ ] Brainstorm what is appropriate for per-project overrides (configured by files outside the project root the sandbox enforces, of course)

### v2.0

* [ ] Brainstorm options for allowing a single command to map to a multi-stage build with each stage having different restrictions. (Might be out-of-scope for this tool and instead handled by some helper tool that invokes this tool separately for `cargo fetch`, `cargo build`, etc.)

**Comment by NobodyXu** [comment]
Cargo has a command called [`cargo locate-project --message-format plain`](https://doc.rust-lang.org/cargo/commands/cargo-locate-project.html) which prints the root of the project to stdout, I think we can use this to find the root of the project.

**Comment by ssokolow** [comment]
That would involve spawning an additional subprocess on each invocation. Do you have evidence that doing it that way is an improvement over just walking up the filesystem ourselves to find `Cargo.toml`?

After all, not all build systems that will need to be supported are going to have an equivalent command and some (eg. `npm`) are horrendously slow to start up, so the machinery to walk up the filesystem looking for a file is going to need to be implemented either way.

**UPDATE:** ...and it would be an additional point of failure, since the command being invoked to identify the path to sandbox could itself become compromised and can't be run sandboxed.

**Comment by NobodyXu** [comment]
The problem with manually finding `Cargo.toml` is workspace.

In that particular situation, finding `Cargo.toml` is not enough,

Edit:

After finding the first `Cargo.toml`, you would also have to check whether its parent dirs has `Cargo.toml` and verify that.

IMHO in the v0.1 MVP we could just invoke an external command.

**Comment by NobodyXu** [comment]
I also would like the following `firejail` sandbox options to be applied when building packages:

```
private-etc alternatives,ca-certificates,ld.so.cache,ld.so.conf,ld.so.conf.d,ld.so.preload,resolv.conf,ssl

private-dev
private-tmp

caps.drop all
nonewprivs # Make setuid binaries has no effect

noroot

dbus-system none
dbus-user none


blacklist /opt
blacklist /media
blacklist /mnt
blacklist /srv
blacklist /sys
blacklist /run

blacklist /boot
blacklist /root
blacklist /sbin
blacklist /snap
blacklist /sys
blacklist /var

noexec /bin/su
noexec /usr/bin/sudo
```

**Comment by ssokolow** [comment]
> After finding the first Cargo.toml, you would also have to check whether its parent dirs has Cargo.toml and verify that.

That was what the `root_find_outermost=true  # For workspaces` in the draft config file was.

I was already taking the need to find the last match rather than the first into account.

> I also would like the following firejail sandbox options to be applied:

I'm pretty paranoid about that, though some of those may have to wait until v1.1 because I'm not sure per-profile Firejail rules will make v1.0 and I suspect that `private-etc` and some of those `blacklist`s might break some of the non-Cargo build tools I want to support.

**Comment by NobodyXu** [comment]
> That was what the root_find_outermost=true # For workspaces in the draft config file was.
> 
> I was already taking the need to find the last match rather than the first into account.

Sounds good to me.

> I'm pretty paranoid about that, though some of those may have to wait until v1.1 because I'm not sure per-profile Firejail rules will make v1.0 and I suspect that private-etc and some of those blacklists might break some of the non-Cargo build tools I want to support.

I think that at least some of them should be applied unconditionally:

```
private-dev
private-tmp

caps.drop all
nonewprivs # Make setuid binaries has no effect

noroot

noexec /bin/su
noexec /usr/bin/sudo

blacklist /sys
blacklist /boot
```

**Comment by NobodyXu** [comment]
I would also like to decide the language to be used for this project.

IMO dynamic-typing or weak-typing languages like python and bash is not suitable for security-sensitive application like this since you can easily work around their typing system.

Since I don't have much experience in Go, IMHO using Rust in this project sounds reasonable.

**Comment by NobodyXu** [comment]
It seems that `firejail` already has a [`cargo.profile`](https://github.com/netblue30/firejail/blob/master/etc/profile-a-l/cargo.profile) though IMO it is actually too strict.

**Comment by ssokolow** [comment]
> I think that at least some of them should be applied unconditionally:

I'm certainly hoping to do more than that for v0.1. I'm just not sure we can do *all* of the longer list until we have support for per-profile overrides.

> Since I don't have much experience in Go, IMHO using Rust in this project sounds reasonable.

I thought using Rust was being taken as a given before we even moved off Reddit. The only reason I even mentioned Python is that, at one point, I was considering starting from a Python script I'd written for something else.

Bash is certainly *not* something I'd want to write something longer than a dozen lines in (I use Python for my quick "shell scripting" and Rust for my more long-term "shell scripting") and I've never used Go because I don't like how primitive it is.

> It seems that `firejail` already has a [cargo.profile](https://github.com/netblue30/firejail/blob/master/etc/profile-a-l/cargo.profile) though IMO it is actually to strict.

I generally only take ready-made Firejail profiles as advice on what I can do in my own from-scratch profiles without causing the program to fail to start.

**Comment by NobodyXu** [comment]
> I'm certainly hoping to do more than that for v0.1. I'm just not sure we can do all of the longer list until we have support for per-profile overrides.

Me too, I also cannot be 100% sure that all of them in the longer list is appliable.

> I thought using Rust was being taken as a given before we even moved off Reddit. The only reason I even mentioned Python is that, at one point, I was considering starting from a Python script I'd written for something else.
> 
> Bash is certainly not something I'd want to write something longer than a dozen lines in (I use Python for my quick "shell scripting" and Rust for my more long-term "shell scripting") and I've never used Go because I don't like how primitive it is.

Cool.

> I generally only take ready-made Firejail profiles as advice on what I can do in my own from-scratch profiles without causing the program to fail to start.

I just thought it can be used as a reference.

**Comment by ssokolow** [comment]
> I just thought it can be used as a reference.

No problem. Thanks for saving me the trouble of going looking.

**Comment by ssokolow** [comment]
I just thought of another thing that'll need to be configurable. Whether to canonicalize (i.e. resolve symlinks) or just "make absolute" when ascending to find the ancestor in the path which contains `root_marked_by`.

Some build tools may do one while some may do the other so, to Just Work™ without introducing confusing edge cases that may cause problems, it'll be necessary to have something like `root_resolve_symlinks` to enable matching what the command to be wrapped does.

(eg. Rust has `std::fs::canonicalize` but no "make absolute" in `std` yet while Python calls "just make absolute" `os.path.abspath` which sounds more inviting and "correct to intention" than `os.path.realpath`, which is how you canonicalize.)

**Comment by NobodyXu** [comment]
> I just thought of another thing that'll need to be configurable. Whether to canonicalize (i.e. resolve symlinks) or just "make absolute" when ascending to find the ancestor in the path which contains `root_marked_by`.
> 
> Some build tools may do one while some may do the other so, to Just Work™ without introducing confusing edge cases that may cause problems, it'll be necessary to have something like `root_resolve_symlinks` to enable matching what the command to be wrapped does.

That's indeed another thing that needs to be configured.

**Comment by ssokolow** [comment]
That said, given that Rust has no "make absolute" in `std` yet, I'll need to decide how to go about it.

(i.e. For security reasons, it may be a good idea to just do a one-to-one port of the Python code until something for that shows up in the Rust standard library rather than relying on a third-party crate.)

**Comment by NobodyXu** [comment]
> (eg. Rust has std::fs::canonicalize but no "make absolute" in std yet while Python calls "just make absolute" os.path.abspath which sounds more inviting and "correct to intention" than os.path.realpath, which is how you canonicalize.)

According to [here](https://stackoverflow.com/questions/30511331/getting-the-absolute-path-from-a-pathbuf), [`std::fs::canonicalize`](https://doc.rust-lang.org/std/fs/fn.canonicalize.html) seems to return an absolute path.

**Comment by ssokolow** [comment]
It's a wrapper around the libc `realpath(3)` function. "realpath() expands all symbolic links".

`os.path.abspath` just performs string manipulation based on `$PWD` and the given path.

They behave differently in a situation like this:

* `~/src/project1/Cargo.toml`
* `~/src/project2/Cargo.toml`
* `~/src/project1/src is a symlink to ~/src/project2/src`

Canonicalizing will turn `~/src/project1/src/mymod` into `~/src/project2/src/mymod` before we ascend looking for `Cargo.toml` while a non-canonicalizing "make absolute" function will not.

**Comment by NobodyXu** [comment]
Thanks for explaining this to me.

I didn't realize that symlink handling is different.

**Comment by ssokolow** [comment]
As for how we do it, since there's no "just make absolute" function in `std` yet, I may want to do a one-to-one port of the code for `os.path.abspath` from Python and then do some differential fuzzing to make sure that the behaviour is identical to the mature Python implementation.

It's a little too central to the security proposition for me to feel comfortable relying on a third-party crate.

**Comment by NobodyXu** [comment]
Agreed.

**Comment by ssokolow** [comment]
Of course, I'll put it in its own module and file so keeping the licenses separate and the ported code easy to replace is as simple as possible.

**Comment by ssokolow** [comment]
Another thought to consider when I have time to do some research. Maybe I *should* use a proper argument parser but configure it in "the first non-option implies `--` (all following arguments are positional)" mode and support options like `--allow-network` to make it feasible to do something like "Run `nodo --allow-network npx <command>` once to install it, then run `nodo npx <command>` to use it".

(The main question being whether there are other details about how commands like `npx` function which would make that not the best choice. For example, how does its cache work?)

**Comment by NobodyXu** [comment]
I am not familiar with npm.

AFAIK about cargo, this seems to be perfectly OK.

**Comment by ssokolow** [comment]
`npx` is a command that comes with NPM that's essentially a hybrid of `cargo install` and `cargo run`. It'll download the package you specify, cache it to a user-global install location akin to `~/.cargo/bin` but not in `PATH`, and then run the command.

Once [WASI](https://wasi.dev/)'s APIs are richer and more crates support targeting it, [`wax`](https://docs.wasmer.io/ecosystem/wapm/wax) would be a Rust equivalent with built-in sandboxing for anything published to Wasmer's [WAPM](https://wapm.io/) repository.

In fact, this project is inspired by `wax`'s default sandboxing behaviour. (Because WASI uses capabilities-based security, you have to pass in handles for the pieces of the filesystem you want the program to be able to see. `wax` defaults to granting `$PWD` when you run something with it.)

**Comment by NobodyXu** [comment]
Sounds amazing, though the combination of downloading and run makes sandboxing it harder.

**Comment by ssokolow** [comment]
You can always use the `wapm` command to download and run things separately. `npx` and `wax` are just wrappers to make things as simple as possible.

Also, WebAssembly has its own sandboxing so, if you trust `wax`, it can deny the command network access without needing to itself be denied permission to access the network for updates. (Like JavaScript running in a browser with uMatrix.)

In fact, `wax` is just a wrapper script around `wapm execute` minus the `--offline` flag. `wapm install <command>` with one sandbox profile, then `wapm execute --offline <command>` with another.

**Comment by NobodyXu** [comment]
IMHO, it’s probably not necessary to sandbox wasm since it has builtin sandbox via its capacity based API.

So maybe we should focus on cargo and npm instead, since these two doesn’t have any builtin sandbox?

**Comment by ssokolow** [comment]
With the way I want this to operate, once we've got profiles for Cargo and `npx` set up, it's just copy-pasting a dozen lines of TOML and editing the command and subcommand names to support WAPM and wax.

The most time-consuming part would be running the tests to Do It Right™ and confirm my suspicion that, being written in Rust, Wasmer's WAPM CLI tool uses `std::fs::canonicalize` for making paths absolute.

**Comment by NobodyXu** [comment]
That matches with my impression of this project so far.

On the other hand, where shall we put the configuration file?
I think `$HOME/.config/nodo/config.toml` will be a good place for such config file.

Also, I think we can configuration file splitting v0.1.x so that users can have `cargo-profile.toml`, `npm-profile.toml` and etc.

**Comment by ssokolow** [comment]
Definitely something compliant with the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), so something starting with `${XDG_CONFIG_HOME:-${HOME}/.config}`.

I'll want to think about whether to re-implement that to protect people who install with `cargo install` without `--locked` though.

I'm not sure about configuration file splitting though. I've learned to follow [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) the hard way and I haven't yet had a solid rationale that convinces me it's worth the added maintenance burden and risk of overlooking a way it could make the design more vulnerable.

That sort of thing is more for when you want your creation to be a piece of infrastructure a project like Cargo or NPM can integrate support for, and that would divide the responsibility for ensuring the configurations are trustworthy compared to just having all the vetted ones in the same repo as the code which enforces them.

...but I will want to think about whether it's worthwhile to let users extend the default config rather than completely overriding it, given that things like adding new profiles would require them to add new `alias`es, and making the rules more nuanced will probably result in a new schema version for the config file.

(I do plan to have a key like `schema_version=1` at the top level which can be used to check when the user should receive a warning about outdated rules and how to revise them.)

**Comment by NobodyXu** [comment]
> I'm not sure about configuration file splitting though. I've learned to follow YAGNI the hard way and I haven't yet had a solid rationale that convinces me it's worth the added maintenance burden and risk of overlooking a way it could make the design more vulnerable.

Maybe we should wait for user feedback.
If the configuration file is simple enough, then splitting it definitely isn't necessary.

> ...but I will want to think about whether it's worthwhile to let users extend the default config rather than completely overriding it, given that things like adding new profiles would require them to add new aliases, and making the rules more nuanced will probably result in a new schema version for the config file.

Agreed, we should also provide a sensible default configuration.

**Comment by ssokolow** [comment]
> Agreed, we should also provide a sensible default configuration.

We should aspire to have 99% of users not needing anything but the defaults.

**Comment by ssokolow** [comment]
Just to get it noted down, the main thing I'll want to sit and think about once I've got a proto-v0.1 to experiment with is the schema for the config file.

With needing to support subcommands (eg. `fetch`) *and* flags (eg. `--offline`) as means to control whether things like network access should be granted, the draft design shouldn't be the final one.

To achieve that properly, I'll probably want to sit down and write out a sheet with all the different commands (eg. `cargo`, `npm`, etc.), subcommands, and flags that are relevant to the behaviour of the sandbox, and then brainstorm on the design which can describe them in the cleanest way.

...either that or just treat schema version 1 as temporary and accept that, for 99% of use-cases, it *should* be expressive enough, even if it's a little inelegant.

**Comment by NobodyXu** [comment]
Maybe we can have a section like “[profile.cargo-fetch]” and override and extend configurations provided in “[profile.cargo]”?

**Comment by NobodyXu** [comment]
I just realised today that Godbolt also uses sandbox internally.

After a quick look at the source code, I found the [firejail configuration they use](https://github.com/compiler-explorer/compiler-explorer/tree/main/etc/firejail).

**Comment by ssokolow** [comment]
> Maybe we can have a section like “[profile.cargo-fetch]” and override and extend configurations provided in “[profile.cargo]”?

I considered that briefly, but it feels like it'll get verbose and/or complicated quickly with the interactions of things like `cargo` defaults (no network), overridden by `fetch` and `build` and `run` (network), overridden by `--offline` (no network) if we don't sit down and carefully plan it out.

That's why I want to research all the commands I can think of supporting first... so I can sit down and carefully plan properly.

(Remember, my goal is a sandboxing thing that Just Works™ if you stick `nodo` onto the beginning of your usual commands, so it has to be expressive enough to handle cases like that.)

**Comment by ssokolow** [comment]
...also, maybe I should give it a `-v` option that causes it to print something like `nodo: profile=cargo +net -gui -dbus root=/path/to/detected/root` as the first line of output as part of a comprehensive plan to steer users away from just blindly dropping the `nodo` if a build fails in the hope that will make it work.

**Comment by NobodyXu** [comment]
> > Maybe we can have a section like “[profile.cargo-fetch]” and override and extend configurations provided in “[profile.cargo]”?
> 
> I considered that briefly, but it feels like it'll get verbose and/or complicated quickly with the interactions of things like `cargo` defaults (no network), overridden by `fetch` and `build` and `run` (network), overridden by `--offline` (no network) if we don't sit down and carefully plan it out.
> 
> That's why I want to research all the commands I can think of supporting first... so I can sit down and carefully plan properly.
> 
> (Remember, my goal is a sandboxing thing that Just Works™ if you stick `nodo` onto the beginning of your usual commands, so it has to be expressive enough to handle cases like that.)

One issue with that is that cargo can be extended…

**Comment by ssokolow** [comment]
Of course, but if you're extending Cargo, then it's perfectly reasonable for you to be expected to extend your sandbox definitions. That's one reason the syntax needs to be simple and easy to work with.

...still, I *do* intend to include ready-made definitions for some of the most common third-party subcommands.

**Comment by NobodyXu** [comment]
> ...still, I do intend to include ready-made definitions for some of the most common third-party subcommands.

Yeah.

> That's why I want to research all the commands I can think of supporting first... so I can sit down and carefully plan properly.

I tried to list alll commands supported by cargo's stable channel and split them into 4 categories:

Commands that requires network access and modifies `$HOME/.cargo`:

```
    install              Install a Rust binary. Default location is $HOME/.cargo/bin
    login                Save an api token from the registry locally. If token is not specified, it will be read from stdin.
    logout               Remove an API token from the registry locally
    search               Search packages in crates.io
    uninstall            Remove a Rust binary
```

Commands that creates new project (need special handling since `Cargo.toml` does not exist):

```
    init                 Create a new cargo package in an existing directory
    new                  Create a new cargo package at <path>
```

Commands that only needs access to the project:

```
    verify-project       Check correctness of crate manifest
    version              Show version information
    check                Check a local package and all of its dependencies for errors
    clean                Remove artifacts that cargo has generated in the past
    clippy               Checks a package to catch common mistakes and improve your Rust code.
    d                    alias: doc
    doc                  Build a package's documentation
    fmt                  Formats all bin and lib files of the current crate using rustfmt.
    locate-project       Print a JSON representation of a Cargo.toml file's location
    metadata             Output the resolved dependencies of a package, the concrete used versions including overrides, in machine-readable format
    pkgid                Print a fully qualified package specification
    read-manifest        Print a JSON representation of a Cargo.toml manifest.
    rustdoc              Build a package's documentation, using specified custom flags.
```

Commands that requires network access, access to the project and `$HOME/.cargo`:

```
    b                    alias: build
    bench                Execute all benchmarks of a local package
    build                Compile a local package and all of its dependencies
    c                    alias: check
    fetch                Fetch dependencies of a package from the network
    fix                  Automatically fix lint warnings reported by rustc
    generate-lockfile    Generate the lockfile for a package
    package              Assemble the local package into a distributable tarball
    owner                Manage the owners of a crate on the registry
    publish              Upload a package to the registry
    r                    alias: run
    run                  Run a binary or example of the local package
    rustc                Compile a package, and pass extra options to the compiler
    t                    alias: test
    test                 Execute all unit and integration tests and build examples of a local package
    tree                 Display a tree visualization of a dependency graph
    yank                 Remove a pushed crate from the index
    update               Update dependencies as recorded in the local lock file
    vendor               Vendor all dependencies for a project locally
```

Since the commands above all build the project, they also might need to download the dependencies or `build.rs` might access the network.

Also, some users might use `sccache` (yeah, I use it) and it can be configured to distribute the compilation jobs and use cache stored in external machines.

**Comment by NobodyXu** [comment]
Here are some cargo extensions that I know:

Commands that requires network access, access to the project and `$HOME/.cargo`:
 - `cargo-criterion`
 - `cargo-expand`
 - `cargo-flamegraph`
 - `cargo-tarpaulin`

They are need to compile the project, so they would also need to download dependencies, which requires network.

**Comment by ssokolow** [comment]
Would you mind running through the commands that require network access to check how many follow the convention of having an `--offline` flag which the sandbox should recognize as a signal to not allow network?

**Comment by NobodyXu** [comment]
> Would you mind running through the commands that require network access to check how many follow the convention of having an `--offline` flag which the sandbox should recognize as a signal to not allow network?

It seems that all the builtin commands of `cargo` that requires network access has `--offline` in their help page.

And then I checked `cargo --help` and found that `--offline` is one of the standard flags that every `cargo` subcommand has.

**Comment by NobodyXu** [comment]
For the cargo extensions I mentioned:
 - `cargo-expand`
 - `cargo-flamegraph`

does not have `--offline`.

**Comment by ssokolow** [comment]
In that case, I suppose the question is whether we should offer a "Just Works, but less secure" mode of operation where any unrecognized subcommand gets network access (unless `--offline` is seen) and whatever solution we come up with for `$HOME/.cargo`.

(eg. I'm thinking that we'll want to force `$HOME/.cargo/bin` to read-only access for every subcommand that has `$HOME/.cargo` access except `install` and `uninstall`.)

**Comment by NobodyXu** [comment]
> eg. I'm thinking that we'll want to force $HOME/.cargo/bin to read-only access for every subcommand that has $HOME/.cargo access except install and uninstall.

I will add that most commands except for `cargo publish` does not access `$HOME/.cargo/credentials`.

And we should also make `$HOME/.cargo/config.toml` read-only.

**Comment by NobodyXu** [comment]
> In that case, I suppose the question is whether we should offer a "Just Works, but less secure" mode of operation where any unrecognized subcommand gets network access (unless --offline is seen) and whatever solution we come up with for $HOME/.cargo.

I think we can implement this in v0.1 MVP, then improve on it in the subsequent revisions.

**Comment by ssokolow** [comment]
Yeah.

Also, for a later v0.1.x, it'd probably be a good idea to do a survey of Cargo subcommands to identify what *inside* the project root could be forced to read-only access for which commands.

(eg. I don't imagine most commands will execute anything like `build.rs` that might need permission to modify the contents of `src`, and we can whitelist the things allowed to modify `Cargo.toml`.)

**Comment by NobodyXu** [comment]
> Yeah.
> 
> Also, for a later v0.1.x, it'd probably be a good idea to do a survey of Cargo subcommands to identify what _inside_ the project root could be forced to read-only access for which commands.
> 
> (eg. I don't imagine most commands will execute anything like `build.rs` that might need permission to modify the contents of `src`, and we can whitelist the things allowed to modify `Cargo.toml`.)

Totally agreed.

Modifying the project other than `target` should be forbidden.

Honestly, most of the `build.rs` don't modify the source.
Binding generator like `bindgen` usually outputs the generated binding to `target`, and the file will be included in a predefined module in `src/` using ~~something like `include_file` (can't remember the name)~~ `include!(concat!(env!("OUT_DIR"), "/binding.rs"));`

Though we also need to take into account that user can configure `cargo` to use a custom "target" using `--target-dir`.

**Comment by NobodyXu** [comment]
Another thing I want to add to the configuration is `whitelist_path`.

Some users like me who use [`sccache`](https://github.com/mozilla/sccache) would like the `cargo build` to be able to access the cache.

**Comment by ssokolow** [comment]
That's why I'm not going for full whitelisting until at *least* 0.2.x. You wind up deep in the weeds reinventing parsing every command's arguments if you try for that.

Well down the curve of diminishing returns. For the most part, I'm just thinking about this stuff now because it will help to reduce how many different revisions the config schema will have to go through to keep up with demands.

**Comment by NobodyXu** [comment]
> That's why I'm not going for full whitelisting until at least 0.2.x. You wind up deep in the weeds reinventing parsing every command's arguments if you try for that.

Yeah indeed.
It's not worth it at this stage of the project.

Edit:

Though I am fairly certain we can make `Cargo.toml`, `LICENSE`, `REAMD.md`, `examples`, `src`, `tests` and `benches` read-only.

No sane `build.rs` would modify these directories.
`build.rs` that modifies these directories are either extremely poorly written and is very rare, or they are straight-up hacked by others.

**Comment by ssokolow** [comment]
...umm, actually, I think I might have written a `build.rs` at one point that wrote to `src` because I couldn't figure out how to get certain things to play nicely together with the officially recommended output directory.

That sort of thing is why I err *very* much on the side of "Make it just work in the face of sloppy legitimate uses" things for early plans. The more I err in that direction, the less configurable the early releases need to be in order to meet design goals for everyone.

**Comment by NobodyXu** [comment]
> ...umm, actually, I think I might have written a build.rs at one point that wrote to src because I couldn't figure out how to get certain things to play nicely together with the officially recommended output directory.

For bindgen, it writes to somewhere in target/ and set the OUT_DIR, which you can include that file using:

    include!(concat!(env!("OUT_DIR"), "/binding.rs"));

> That sort of thing is why I err very much on the side of "Make it just work in the face of sloppy legitimate uses" things for early plans. The more I err in that direction, the less configurable the early releases need to be in order to meet design goals for everyone.

I agree with it, but I really think for build.rs we should enforce stricter rules.

According to [the official doc](https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script):

> Build scripts may save any output files in the directory specified in the OUT_DIR environment variable. Scripts should not modify any files outside of that directory.

If the build.rs doesn’t play by the rules, then it should be fixed.

**Comment by ssokolow** [comment]
> we should enforce

This phrase is the problem. We're not in a position to make demands of the end-users we service or they'll consider it too onerous for the tool to achieve the popularity I want of it.

The purpose of the tool, first and foremost, is to improve the baseline. Beyond that, we'll see how things go later.

That means we *facilitate* users improving the security of what they're already doing for now.

**Comment by NobodyXu** [comment]
> > we should enforce
> 
> This phrase is the problem. We're not in a position to make demands of the end-users we service or they'll consider it too onerous for the tool to achieve the popularity I want of it.
> 
> The purpose of the tool, first and foremost, is to improve the baseline. Beyond that, we'll see how things go later.

OK, that’s totally reasonable.

**Comment by NobodyXu** [comment]
On a second though, we should makes $HOME/.rustup read-only for all cargo operations.

There is absolutely no way for cargo to modify that directory.

**Comment by ssokolow** [comment]
That makes sense.

* Everything outside the project directory should be inaccessible unless a need is demonstrated otherwise. (With the exception of things like `/usr` where we follow Firejail's lead on trusting the OS permissions for our "retrofit an improved baseline" approach, given how many projects may want to use some APT/RPM/etc.-installed tools or libraries.)
* Everything that's demonstrated to be necessary should be read-only unless a need is demonstrated otherwise.

**Comment by NobodyXu** [comment]
> That makes sense.
> 
> * Everything outside the project directory should be inaccessible unless a need is demonstrated otherwise. (With the exception of things like `/usr` where we follow Firejail's lead on trusting the OS permissions for our "retrofit an improved baseline" approach, given how many projects may want to use some APT/RPM/etc.-installed tools or libraries.)
> * Everything that's demonstrated to be necessary should be read-only unless a need is demonstrated otherwise.

I agree with you.

IMO the only directory outside project which we should consider exposing is $HOME/.cargo and $HOME/.rustup.

The exception is “cargo run”/“npm run”, where user might want additional directory exposed.

**Comment by ssokolow** [comment]
> The exception is “cargo run”/“npm run”, where user might want additional directory exposed.

Users can always alter the configuration as needed, so we don't need to think about that.

**Comment by NobodyXu** [comment]
> Users can always alter the configuration as needed, so we don't need to think about that.

That’s good.

**Comment by NobodyXu** [comment]
@ssokolow I happen to be free recently, so why not start development now?

**Comment by ssokolow** [comment]
I'm trying to focus on getting things cleaned up for Christmas, but getting the features for v0.1 implemented without commiting to even the minimal API stability an MVP leading in to v0.1.1 might imply should be pretty quick.

I'll see if I can fit that in within the next few days.

**Comment by ssokolow** [comment]
Something that just occurred to me.

If it feels like it can be represented comfortably enough in the TOML, maybe, instead of `*_subcommands` fields, the `CommandProfile` struct should contain a `subcommands: Vec<String, CommandProfile>` field and things like "deny network by default but allow for these subcommands" would just be a "deny network" setting on the parent command and an "allow network" setting on each subcommand, with the subcommands literally just being command definitions that are in the `subcommands` field of another command definition, instead of in the top-level profile list.

The main two potential issues I see are:

1. Duplication for things where a subcommand has aliases or multiple subcommands need the same permissions
2. Whether something like `[profiles.cargo.subcommands.fetch]` and `[profiles.npm.subcommands.install.subcommands."-g"]` is the best I can do for user comfort in that concept.

\*shrug* Something to think about after I go to bed.

**Comment by NobodyXu** [comment]
Maybe we can have something like this:

    cmd_profile = [
        “b”: “build”,
        “builld”: […],
     ]

I am not very familiar with toml, so I could be wrong on the syntax.

**Comment by ssokolow** [comment]
> Maybe we can have something like this:

The problem is that now you've got a schema which is hostile to commands that don't *have* subcommands or where all subcommands should use the same set of sandboxing rules.

If it were XML, I'd suggest using the availability of different tag names so rules and subcommands can exist at the same level without getting confused but, with JSON, YAML, or TOML, that's not an option. 

>  I am not very familiar with toml, so I could be wrong on the syntax.

Here's a TOML<->JSON converter you can experiment with: https://pseitz.github.io/toml-to-json-online-converter/

...and here's the TOML spec: https://toml.io/en/

TOML is best when you design your schema so it can feel like a more powerful variant of `.ini` files.

**Comment by NobodyXu** [comment]
Thanks, I will take a look at it.

**Comment by NobodyXu** [comment]
> The problem is that now you've got a schema which is hostile to commands that don't have subcommands or where all subcommands should use the same set of sandboxing rules.

Perhaps we can hardcode the aliases.

AFAIK, these aliases are for builtin commands only and is fairly limited.

Hard coding them might not be a problem at all and would make the configuration file simpler.

**Comment by ssokolow** [comment]
As far as I'm concerned, it's unacceptable to hard-code *anything* specific to any subprocess other than `firejail` beyond baking a default config file into the binary using `include_str!` which can be dumped and edited using something like `--dump-config`.

Cargo Workspaces need to be supported? Invent the `root_find_outermost=true` config key. `make` only checks the current directory for `Makefile`? Invent the `cwd_to_root=true` config key. etc.

Heck, I'm trying to be careful about how specific I am for the permissions so that, if a backend other than Firejail gets supported in the future, there will be minimal disruption involved.

**Comment by NobodyXu** [comment]
Sorry for that, I didn't take into account of that.

Maybe we should have another variable handling the aliases, e.g.:

    let aliases = [
        "b": "build",
    ]

**Comment by ssokolow** [comment]
That *is* a good idea and probably along the lines I'd have come up with.

In TOML, it could use either the inline form for cases with few entries:

```toml
[profile.cargo]
subcommand_aliases = {r="run", t="test", b="build"}
```

...or the table form for cases with many entries:

```toml
[profile.cargo.subcommand_aliases]
r="run"
t="test"
b="build"
```

**Comment by ssokolow** [comment]
It looks like I might have to push this to the new year. Christmas-related preparation is taking longer than expected and I promised myself I'd stick to leisure activities I've been neglecting for two weeks starting on Christmas.

**Comment by NobodyXu** [comment]
@ssokolow Totally understandable!

I also want to have some fun and spend more time with my friends rather than sitting at home all day, since I have been under lockdown for at least 4 months in Sydney.

Have a good Christmas!

We can talk about this in 2022.

**Comment by ssokolow** [comment]
Sorry for taking so long. I've been working on the configuration side of the project over the last couple of days to get that hammered down and I realized that `cwd_to_root` should be removed as a feature.

Specifically, because changing the working directory will break things like `make -f NameOtherThanMakefile` and it just presents too many opportunities for surprising behaviour that could open up a security vulnerability.

**Comment by NobodyXu** [comment]
I am good with removing `cwd_to_root`.

This would actually make it easier to implement the initial version.

**Comment by ssokolow** [comment]
How does this look as a draft for a default "should work for most people" set of Firejail flags shared between all profiles?

```toml
# The set of Firejail flags that get applied to *all* profiles
#
# As flags in this list don't yet have more specific preferences which control
# them, feel free to change this if you're working on projects which have
# specialized needs. For example:
#
# - If you store your projects on removable media, remove `--blacklist=/media`
# - If you need to play sounds, remove both `--nosound` and `--noautopulse`
# - If you need access to serial ports, remove `--nogroups` since that is
#   typically achieved via membership in the `dialout` group.
# - If you need to display a GUI, remove `--no3d`, since
#   modern GUI toolkits tend to use at least some degree of GPU acceleration.
#
# Conversely, if you feel these "Just Work™ in most cases" defaults aren't
# strict enough, feel free to add flags to lock things down further.
#
# NOTE: `--x11=none` cannot be included here by default because Firejail will
#       refuse to start if you specify it without either setting your X server
#       configuration to `-nolisten local` or passing Firejail a suitable
#       `--net` flag, which is a per-profile choice.
firejail_base_flags=[
    "--blacklist=/media",
    "--blacklist=/mnt",
    "--blacklist=/srv",
    "--caps.drop=all",
    "--deterministic-exit-code",
    "--no3d",
    "--nogroups",
    "--nonewprivs",
    "--noroot",
    "--nosound",
    "--noautopulse",
    "--novideo",
    "--nou2f",
    "--private-dev",
    "--private-tmp",
    "--protocol=unix,inet,inet6,netlink",
    "--seccomp",
    "--shell=none",
    "--quiet",
]
```

(Bear in mind that `--net=none` and `--whitelist` and `--blacklist` flags will be provided by a mixture of the profile and dynamically generated rules like "don't let sandboxed apps mess with the sandboxing configuration file".)

**Comment by NobodyXu** [comment]
It looks good to me.

I think maybe it is better to also disable dbus access via `--dbus-system=none` and also blacklist `/run/mount` and `/run/media`?

Since we created a mount namespace, maybe we should also enable IPC namespace via `--ipc-namespace`?

**Comment by ssokolow** [comment]
> I think maybe it is better to also disable dbus access via `--dbus-system=none`

`--dbus-system` isn't present in the version of Firejail on my Ubuntu 20.04 LTS system and I'd prefer to avoid having to do version detection.

As for `--nodbus` as an alternative, I don't want to give users a false sense of security and I haven't yet figured out how to get the "Parent pid 3156434, child pid 3156435"-silencing effect of `--quiet` without also silencing the "Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set" warning.

> and also blacklist `/run/mount` and `/run/media`?

Where are those specified? I've got `/run/mount` but not `/run/media`, and it seems questionable to blacklist `/run/mount/utab` when I'm already blacklisting the mountpoints it lists and not blacklisting `/etc/fstab`.

(Remember, this is a set of restrictions that may get *looser* after beta testing, because it's intended as a *baseline* that the average user won't need to modify.)

Heck, the main reason I blacklist `/mnt`, `/media`, and `/srv` is that I've got things in those folders that are writable by the user I run as and firejail's `--whitelist` doesn't block them out by default.

> Since we created a mount namespace, maybe we should also enable IPC namespace via `--ipc-namespace`?

Good point. I forgot about that and then missed it during my re-read of `--help` when I skipped past the options related to manipulating IPv4 and IPv6 more finely than using or not using `--net=none`.

**Comment by NobodyXu** [comment]
> As for --nodbus as an alternative, I don't want to give users a false sense of security and I haven't yet figured out how to get the "Parent pid 3156434, child pid 3156435"-silencing effect of --quiet without also silencing the "Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set" warning.

Ok then, let's leave `dbus` options behind for the initial release.

> Where are those specified?

According to [here](https://unix.stackexchange.com/questions/176790/what-is-the-distinction-between-media-mnt-and-run-mount), `udisks` automatically mount drivers/devices to `/run/media`, so I think we should blacklist it to prevent access.

As for `/run/mount`, it stores information of the `mount` performed, including the user that mounted it according to [here](https://unix.stackexchange.com/questions/406256/what-is-run-mount-utab).

I think we should also blacklist it to prevent program from accessing it.

**Comment by ssokolow** [comment]
> According to here, udisks automatically mount drivers/devices to `/run/media`, so I think we should blacklist it to prevent access.

Sounds good. Will do.

> I think we should also blacklist it to prevent program from accessing it.

Again, what is the rationale for blacklisting `/run/mount`?

1. We're not blacklisting `/etc/fstab`
2. We're already blacklisting the partitions `/run/mount` points to
3. This is a general-purpose set of defaults that should Just Work™ as a baseline for most people.
4. The main reason I'm blacklisting those paths is because they're likely to contain writable stuff. (i.e. The baseline isn't generally *intended* to whitelist *read* access to stuff outside `$HOME`.)

**Comment by NobodyXu** [comment]
OK, I think it is reasonable to not blacklist `/run/mount`.

**Comment by ssokolow** [comment]
Anyway, it's late here, so I'll decide whether to iterate on how that is represented in the config file and the schema tomorrow.

(I spent today converting the existing config schema over to a bunch of newtypes, replacing the bools with custom enums, and writing unit tests to make it as difficult as possible for later stages to confuse things and possibly open up a security hole.)

For example, while it's generated by a helper macro, this is rustdoc's view of what the `allow_network=true|false` profile field parses into:

```rust
pub enum Network {
    ChildProcsOnly,
    AllNetworks,
}
```

**EDIT:** Here it is in context:

![screenshot](https://user-images.githubusercontent.com/46915/151296872-ff84aae0-ee26-40d2-841b-095a2bec3a45.png)

**Comment by NobodyXu** [comment]
It looks quite good.

Can you push it to the repository though?

It will be much easier to understand the current design with the source code.

**Comment by ssokolow** [comment]
My main concern is that I haven't yet gone through and made sure all my relevant accounts are as secure as I can make them and, once this is more than just a discussion thread, the clock is potentially ticking for me to become a target for supply-chain attacks.

However, since it's just the config parsing so far, I suppose there's no harm in that.

**Comment by NobodyXu** [comment]
It is understandable to have concern over supply-chain attacks, though I don't think there will be any people targeting this software at this stage.

**Comment by ssokolow** [comment]
There we go. Done.

...though I intentionally followed the reverse of GitHub's advice for the short term:

> Help people interested in this repository [not find] your project by [not] adding a README.

**Comment by ssokolow** [comment]
Oh, don't hesitate to use `cargo doc --open` to get a feel for the codebase. I'm pretty proactive about polishing up my API documentation for easy reading.

Ideally, I'd also like to have this complete and polished enough to be using it to harden my own system before I start drawing attention to its existence.

**Comment by ssokolow** [comment]
Another "Oh, ...", feel free to let me know what you think of what I've got so far.

I do my best, and I already had over a decade of experience with other languages when Rust 1.0 hit, but I don't usually write Rust for other people to see and I have a bad habit of *not* finding time to read other people's codebases to pick up tips.

**Comment by NobodyXu** [comment]
I've currently reading the source code.

Once I finished reading, I will comment on it and maybe I will even submit a PR.

Edit:

I will open another issue for discussion on the implementation.

**Comment by ssokolow** [comment]
I pushed a first draft of the argument parsing. Do you have any feedback on the `--help` output?

**Comment by NobodyXu** [comment]
I think the help message for `--` can be improved:

```
    --              All arguments after '--' is treated as part of the command to be executed inside firejail
```

**Comment by ssokolow** [comment]
The problem with that phrasing is that it could lead a user to doubt their understanding of how arguments are parsed because, in most cases, options can be mixed in among positional arguments.

(i.e. They could mistakenly assume that there are cases other than "`<command>` starts with `-` or `--`" where `--` is necessary.)

I'll see if I can come up with a better alternative to my draft that doesn't have that problem.

**Comment by ssokolow** [comment]
OK. How does the revised version look?

**Comment by NobodyXu** [comment]
It is much better and is easy to understand.

**Comment by ssokolow** [comment]
Let me know what you think of the "look up the path to the configuration file" code I just pushed.

**Comment by NobodyXu** [comment]
Looks all good to me, except the use of deprecated `env::home_dir`.

**Comment by NobodyXu** [comment]
Maybe we can use [`dirs::home_dir`](https://docs.rs/dirs/latest/dirs/fn.home_dir.html) or [`dirs_next::home_dir`](https://docs.rs/dirs-next/latest/dirs_next/fn.home_dir.html).

**Comment by ssokolow** [comment]
As it says in the code comment:
1. `env::home_dir` is deprecated because of its behaviour on Windows
2. This is a tool that's fundamentally cgroups-specific and, thus, Linux-specific
3. I don't want to add another external dependency and the associated rise in opportunity for supply-chain attacks.
4. In this niche specifically, the recommended crate already suffered from some drama surrounding how the maintainer chose to act on feeling unable to continue maintaining it and that prompted [rust-lang/rust#71684](https://github.com/rust-lang/rust/issues/71684).
5. The behaviour listed under "Unix" in the rustdoc page is perfectly acceptable.

**TL;DR:** I chose it with full awareness of *why* it's deprecated and having been around to see that the third-party alternatives have already been more iffy than usual.

In this case where I'm writing a security tool, I trust a function that's been deprecated for behaving too Unixy on Windows but is still subject to the Rust 1.0 stability promise as part of `std` over a third-party crate.

**EDIT:** If I didn't consider the deprecated function an acceptable option, I'd go straight for `env::var_os("HOME")` and sacrifice the `getpwuid_r` fallback to keep the lack of an external dependency.

**Comment by NobodyXu** [comment]
Thanks for the explanation, now I think the use of `env::home_dir` is reasonable.

**Comment by NobodyXu** [comment]
Pinging this issue

**Comment by ssokolow** [comment]
Thanks. I switched to some other stuff in my life (including being most of the way to finally doing the decade-delayed configuration rework for my QuickTile project), but then wound up having sleep issues.

I'll probably return to things in reverse order. (Fix sleep, then complete and release the QuickTile rework, then get back to nodo)

**Comment by NobodyXu** [comment]
> Thanks. I switched to some other stuff in my life (including being most of the way to finally doing the decade-delayed configuration rework for my QuickTile project), but then wound up having sleep issues.
> 
> I'll probably return to things in reverse order. (Fix sleep, then complete and release the QuickTile rework, then get back to nodo)

Take a rest and not pressure yourself too much!
I found melatonin to be very helpful when it comes to sleep issues.

**Comment by ssokolow** [comment]
Sorry for going silent. I hadn't realized what a downward spiral I was on, productivity-wise.

I'm currently working to bring all my GitHub repos up to what I consider current standard (Dependabot, Actions, set up cargo-deny, migrations like StructOpt→clap and error-chain→anyhow, etc.) but, once that's done, I'll get back to this.

...and, once this is complete enough for things like the `todo!` usage to be a screaming wall of Fs, I'll add the rust-clippy and https://securityscorecards.dev/ actions to this repo and consider that part of a higher tier of "current standard" for a security-sensitive repo.

**Comment by ssokolow** [comment]
Sorry for taking so long to get back to you.

It took me a while to realize it, but I've become very averse to working on this project because I realized that, if I publish a completed version, and it becomes popular, then, instead of protecting me from supply chain attacks on other crates, it increases the risk that it'll make me an attractive target for *direct* attacks on *this* crate.

I'm still deciding what to do about that. Maybe implement everything short of the final integration, hand the repo ownership over to you, and add a note in the repo to make it clear that I *don't* have commit access anymore and hacking me won't benefit attackers?

**Comment by NobodyXu** [comment]
> Maybe implement everything short of the final integration, hand the repo ownership over to you, and add a note in the repo to make it clear that I don't have commit access anymore and hacking me won't benefit attackers?

Yes, I'm ok with that and I'm willing to help you with the implementation.
I think moving this repo to an org and have strict access control (requires approval) and setup CI for automatic publishing (no maintainer has the right to directly publish anything) makes sense if you are worried about attacks like that.

**Comment by ssokolow** [comment]
OK. November is very busy for me, but I'll try to get back into the swing of things in early December as part of cleaning/catching back up to all the juggling balls I slowly dropped during and in the wake of the pandemic without recognizing the larger pattern.
